datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -710,6 +712,10 @@ private void handleAssignmentChange(boolean isDatastreamUpdate) throws TimeoutEx
       retryHandleAssignmentChange(isDatastreamUpdate);
     }
 
+    if (_config.getEnableAssignmentTokens()) {
+      maybeClaimAssignmentTokensForStoppingStreams(newAssignment, oldAssignment);
```

Comment:
This does not look right. The assignment token must be claimed after all the assigned tasks are stopped. Claiming after submission without knowing that the task has stopped will have a rippling effect. The leader will mark the datastream "STOPPED" and attempt to "RESUME" immediately. With low level, it will result in the same partition getting processed by two tasks, as the task dependency is not maintained outside the restart boundry.

Reply:
Vaibhav, trying to understand your comment here. The follower instance performs this claim after all the assignment futures are waited for, so wouldn't that make sure that the corresponding unassigned tasks in this assignment are stopped within that timeout?

There are a few failure scenarios to think about,
1. If the futures are timing out, we are retrying endlessly, which won't let the instance claim the tokens. 
2. If there's an interrupted exception, we need to handle whether we need the hosts to claim the tokens or not. 

Another optimization may be to think about is,
In case when multiple streams are stopped, and a task for one particular stream cannot be stopped, should we handle the futures per stream/prefix to not block other stopping streams?

Reply:
I checked with Jhora, if that's not the case. Completable Future only confirms that the task state is changed to shutdown. It does not check if the task thread has died or not. The task thread must die to avoid duplicate processing. ([Code block](https://github.com/linkedin/brooklin/blob/2ab7ce0065058913908660b576d93758a0b59594/datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java#L770))